### PR TITLE
New Data Source: `azurerm_storage_containers`

### LINKED ISSUE #24057: Support for getting a list of container names in a storage account
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

There is already an existing datasource `azurerm_storage_container` that requires a container name.  I would like a new datasource azurerm_storage_containerS ? that would give me a list of all of the containers in a named storage account.

It would be fine by me if datasource `azurerm_storage_account` was enhanced to include a list of the container names that it houses.

Bonus: would be to have a filter to only get the names of certain container names that match the filter or regex etc.

The issue we are trying to resolve is that there is an external entity creating containers and we need to be able to get a list of the existing containers and setup IAM and storage management policies.

### New or Affected Resource(s)/Data Source(s)

azurerm_storage_container azurerm_storage_account

### Potential Terraform Configuration

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by magodo**
Fix #24057

## Test

```shell
terraform-provider-azurerm on  storage_container_ds_containers via 🐹 v1.21.4
💤  TF_ACC=1 go test -v -timeout=20h ./internal/services/storage -run='TestAccDataSourcestorageContainers_'
=== RUN   TestAccDataSourcestorageContainers_basic
=== PAUSE TestAccDataSourcestorageContainers_basic
=== RUN   TestAccDataSourcestorageContainers_prefix
=== PAUSE TestAccDataSourcestorageContainers_prefix
=== CONT  TestAccDataSourcestorageContainers_basic
=== CONT  TestAccDataSourcestorageContainers_prefix
--- PASS: TestAccDataSourcestorageContainers_prefix (162.67s)
--- PASS: TestAccDataSourcestorageContainers_basic (163.37s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/storage       163.387s
```

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
@magodo rather than adding this into the `azurerm_storage_account` resource - this is probably a good candidate for a plural `azurerm_storage_containers` data source - which would then allow us to support filtering options on this in the future - for example:

```terraform
data "azurerm_storager_containers" "sources" {
  storage_account_id = "..."
  name_prefix = "example-*"

  # exporting
  #containers {
  #  name = ""
  #  data_plane_id = ""
  #  resource_manager_id = ""
  #}
}
```

In addition since both the Data Plane and Resource Manager IDs are useful depending on the output, I think it'd probably be prudent to export both?

---

An aside, but I think worth mentioning: I'm unsure how useful these'd be in the short-term - longer-term [the `ListCompleteMatchingPredicate` functions](https://github.com/hashicorp/go-azure-sdk/blob/16940698c5a4bc21dd62c81f7d105b8ab700924b/resource-manager/storage/2023-01-01/blobcontainers/method_list.go#L106) are intended to make these "plural" data sources easier to build (particularly in an automated fashion), so using a "plural" data source would align with that.

**Comment by magodo** [comment]
@tombuildsstuff I've reverted the previous change and created the new data source per your comment, please take another look!

File: internal/services/storage/storage_containers_data_source.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Storage.ResourceManager.BlobContainers
+
+			var plan storageContainersDataSourceModel
+			if err := metadata.Decode(&plan); err != nil {
+				return fmt.Errorf("decoding %+v", err)
+			}
+
+			// Ignore the error as is guaranteed by the schema
+			id, _ := commonids.ParseStorageAccountID(plan.StorageAccountId)
```
Comment (tombuildsstuff) [inline internal/services/storage/storage_containers_data_source.go:94]: Can we always raise these regardless? There's been instances where a merge conflict has unintentionally removed validation checks - so this is a good safety net

```suggestion
			id, err := commonids.ParseStorageAccountID(plan.StorageAccountId)
			if err != nil {
				return err
			}
```

File: internal/services/storage/storage_containers_data_source_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/acceptance"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/acceptance/check"
+)
+
+type storageContainersDataSource struct{}
+
+func TestAccDataSourcestorageContainers_basic(t *testing.T) {
```
Comment (tombuildsstuff) [inline internal/services/storage/storage_containers_data_source_test.go:14]: _nit_

```suggestion
func TestAccDataSourceStorageContainers_basic(t *testing.T) {
```

File: internal/services/storage/storage_containers_data_source_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+				),
+				check.That(data.ResourceName).Key("containers.1.data_plane_id").HasValue(
+					fmt.Sprintf("https://acctestacc%s.blob.core.windows.net/test2", data.RandomString),
+				),
+			),
+		},
+	})
+}
+
+func TestAccDataSourcestorageContainers_prefix(t *testing.T) {
```
Comment (tombuildsstuff) [inline internal/services/storage/storage_containers_data_source_test.go:44]:
```suggestion
func TestAccDataSourceStorageContainers_prefix(t *testing.T) {
```

File: website/docs/d/storage_containers.html.markdown [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `storage_account_id` - (Required) The ID of the Storage Account that the Storage Containers reside in.
+
+---
+
+* `name_prefix` - (Optional) A prefix match used for the Storage Container `name` field, case sensitive.
```
Comment (tombuildsstuff) [inline website/docs/d/storage_containers.html.markdown:33]: Terraform is case-sensitive by default (like most systems are), so I don't think we need to call this one out explicitly?

```suggestion
* `name_prefix` - (Optional) A prefix match used for the Storage Container `name` field.
```

File: website/docs/d/storage_containers.html.markdown [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+subcategory: "Storage"
+layout: "azurerm"
+page_title: "Azure Resource Manager: Data Source: azurerm_storage_containers"
+description: |-
+  Gets information about an existing Storage Containers.
+---
+
+# Data Source: azurerm_storage_containers
+
+Use this data source to access information about an existing Storage Containers.
```
Comment (tombuildsstuff) [inline website/docs/d/storage_containers.html.markdown:11]:
```suggestion
Use this data source to access information about the Storage Containers available within a Storage Account.
```

**Review by tombuildsstuff — APPROVED** [review]
A few minor comments but this otherwise LGTM 👍

**Pull request merged**